### PR TITLE
Generalize solarpack code

### PR DESCRIPTION
**Pull request by ymber**
#### Summary
SUMMARY: Infrastructure "Generalize solarpack code"

#### Purpose of change
Allow new items to be defined without engine changes.

#### Describe the solution
Remove hardcoded item IDs from the code and define solar panel efficiency in JSON.

#### Testing
Items behave as before in game. No user facing changes.

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -2366,6 +2366,18 @@ bool Character::worn_with_flag( const std::string &flag, body_part bp ) const
     } );
 }
 
+const item Character::item_worn_with_flag( const std::string &flag, body_part bp ) const
```
Comment (BevapDin) [inline src/character.cpp]: Please don't return a const value, that has almost no usage.

File: src/bionics.cpp [hunk 2]
```diff
@@ -887,14 +887,7 @@ bool Character::burn_fuel( int b, bool start )
...
             if( remote_fuel == fuel_type_sun_light ) {
-                // basic solar panel produces 50W = 1 charge/20_seconds = 180 charges/hour(3600)
-                if( is_wearing( "solarpack_on" ) ) {
-                    effective_efficiency = 0.05;
-                }
-                // quantum solar backpack = solar panel x6
-                if( is_wearing( "q_solarpack_on" ) ) {
-                    effective_efficiency = 0.3;
-                }
+                effective_efficiency = item_worn_with_flag( "SOLARPACK_ON" ).type->armor->solar_efficiency;
```
Comment (BevapDin) [inline src/bionics.cpp]: `item_worn_with_flag` will return a null item when there is no suitable worn item. A null item has no armor data. Accessing `type->armor` will fail, as `armor` would be a null pointer. You need to check that the returned object is actually an armor item.

File: src/item.cpp [hunk 3]
```diff
@@ -8932,8 +8932,7 @@ bool item::process_cable( player *carrier, const tripoint &pos )
     }
     std::string state = get_var( "state" );
     if( state == "solar_pack_link" || state == "solar_pack" ) {
-        if( !carrier->has_item( *this ) || !( carrier->is_wearing( "solarpack_on" ) ||
-                                              carrier->is_wearing( "q_solarpack_on" ) ) ) {
+        if( !carrier->has_item( *this ) || !( carrier->worn_with_flag( "SOLARPACK_ON" ) ) ) {
```
Comment (BevapDin) [inline src/item.cpp]:
```suggestion
        if( !carrier->has_item( *this ) || !carrier->worn_with_flag( "SOLARPACK_ON" ) ) {
```

**Comment by ymber** [comment]
Moved `solar_efficiency` into `itype` so non-wearables can work as solar chargers in future.

**Pull request merged**